oak/server/rust/oak_glue/src/lib.rs

```diff
@@ -0,0 +1,334 @@
+
+#[no_mangle]
+pub extern "C" fn glue_init() {
+    let _ = ::std::panic::catch_unwind(|| {
+        simple_logger::init().expect("failed to initialize logger");
+        info!("Rust FFI glue initialized");
+    });
+}
+
+type NodeFactory =
+    extern "C" fn(data: usize, name: *const u8, name_len: u32, node_id: u64, handle: u64) -> ();
+
+struct Glue {
+    runtime: Arc<oak_runtime::Runtime>,
+    factory: Option<NodeFactory>,
+    factory_data: usize,
+}
+
+impl Glue {
+    fn new(
+        runtime: Arc<oak_runtime::Runtime>,
+        factory: Option<NodeFactory>,
+        factory_data: usize,
+    ) -> Self {
+        Glue {
+            runtime,
+            factory,
+            factory_data,
+        }
+    }
+    /// Recreate a RuntimeProxy instance that corresponds to the given node ID
+    /// value.
+    fn proxy(&self, node_id: u64) -> RuntimeProxy {
+        RuntimeProxy {
+            runtime: self.runtime.clone(),
+            node_id: NodeId(node_id),
+        }
+    }
+}
+
+lazy_static! {
+    static ref GLUE: RwLock<Option<Glue>> = RwLock::new(None);
+}
+
+const R1: &str = "global glue lock poisoned";
+const R2: &str = "global glue object missing";
+
+/// Start the Rust runtime, with the ApplicationConfiguration provided in
+/// serialized form.
+///
+/// # Safety
+///
+/// Caller must ensure that the memory range [config_buf, config_buf+config_len) is
+/// accessible and holds a protobuf-serialized ApplicationConfiguration message.
+#[no_mangle]
+pub unsafe extern "C" fn glue_start(
+    config_buf: *const u8,
+    config_len: u32,
+    factory: Option<NodeFactory>,
+    factory_data: usize,
+) -> u64 {
+    std::panic::catch_unwind(|| {
+        let config_data = std::slice::from_raw_parts(config_buf, config_len as usize);
+
+        let app_config = match ApplicationConfiguration::decode(config_data) {
+            Ok(c) => c,
+            Err(e) => {
+                error!("Failed to decode ApplicationConfiguration: {}", e);
+                return oak_abi::INVALID_HANDLE;
+            }
+        };
+        info!(
+            "start runtime with initial config {}.{}",
+            app_config.initial_node_config_name, app_config.initial_entrypoint_name
+        );
+        let (runtime, grpc_handle) = match oak_runtime::configure_and_run(app_config) {
+            Ok(p) => p,
+            Err(status) => {
+                error!("Failed to start runtime: {:?}", status);
+                return oak_abi::INVALID_HANDLE;
+            }
+        };
+
+        oak_runtime::node::external::register_factory(create_and_run_node);
+        info!("register oak_glue::create_and_run_node() as node factory");
+
+        let glue = Glue::new(runtime, factory, factory_data);
+        *GLUE.write().expect(R1) = Some(glue);
+        grpc_handle.0
+    })
+    .unwrap_or(oak_abi::INVALID_HANDLE)
+}
+
+#[no_mangle]
+pub extern "C" fn glue_stop() {
+    let runtime = GLUE.read().expect(R1).as_ref().expect(R2).runtime.clone();
```

Comment:
May as well get a write lock on the `Option` from the start, and then call `take` on it? It may also avoid the `clone` call (not 100% sure).